File: trapping-rain-water/trapping-rain-water.cc
#include <cstdio>
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
  /*
  int trap(int A[], int n) {
    if(n < 3) return 0;
    int a = 0, b = n-1, leftmax = 0, rightmax = 0, res = 0;
    while(a < b) {
      leftmax = max(leftmax, A[a]);
      rightmax = max(rightmax, A[b]);
      if(leftmax < rightmax) res += (leftmax-A[a]), ++a;
      else res += (rightmax-A[b]), --b;
    }
    return res;
  }
  */
  int trap(int A[], int n) {
    if(n < 3) return 0;
    vector<int> left(n, 0), right(n, 0);
    left[0] = A[0];
    int mmax = A[0];
    for(int i=1; i<n-1; i++) mmax = max(A[i], mmax), left[i] = mmax;
    right[n-1] = A[n-1];
    mmax = A[n-1];
    for(int i=n-2; i>0; i--) mmax = max(A[i], mmax), right[i] = mmax;
    int res = 0;
    for(int i=1; i<n-1; i++) {
      int x = min(left[i], right[i]) - A[i];
      if(x > 0) res += x;
    }
    return res;
  }
};

int main(int argc, const char *argv[])
{
  int A[] = {2,0,2};
  Solution so;
  cout << so.trap(A, sizeof(A)/sizeof(int)) << endl;
  return 0;

}

File: rs/src/quests/trapping_rain_water.rs
/**
 * [42] Trapping Rain Water
 *
 * Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.
 
Example 1:
Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6
Explanation: The above elevation map (black section) is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped.

Example 2:
Input: height = [4,2,0,3,2,5]
Output: 9

 
Constraints:

	n == height.length
	1 <= n <= 2 * 104
	0 <= height[i] <= 105

 */
pub struct Solution {}

// submission codes start here

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let n = height.len();
        if n < 3 { return 0; }
        let (mut left_max, mut right_max) = (height[0], height[n-1]);
        let (mut left, mut right) = (vec![left_max; n], vec![right_max; n]);
        use std::cmp::{min, max};
        (1..n).zip((0..n-1).rev())
            .fold((height[0], height[n-1]), |(left_max, right_max), (i, j)| {
               ({ left[i] = max(left_max, height[i]); left[i] },
                { right[j] = max(right_max, height[j]); right[j] })
            });
        // for (i, j) in (1..n).zip((0..n-1).rev()) {
        //     left_max = max(left_max, height[i]);
        //     left[i] = left_max;
        //     right_max = max(right_max, height[j]);
        //     right[j] = right_max;
        // }
        (0..n).into_iter().fold(0, |acc, i| acc + min(left[i], right[i]) - height[i])
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        assert_eq!(6, Solution::trap(vec![0,1,0,2,1,0,1,3,2,1,2,1]));
        assert_eq!(9, Solution::trap(vec![4,2,0,3,2,5]));
    }
}
